api/test/e2e/upstream_chash_query_string_arg_xxx_test.go

```diff
@@ -0,0 +1,210 @@
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestUpstream_chash_query_string(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "create chash upstream with key (query_string)",
+			Object:   ManagerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/upstreams/1",
+			Body: `{
+					"nodes": [{
+						"host": "172.16.238.20",
+						"port": 1980,
+						"weight": 1
+					},
+					{
+						"host": "172.16.238.20",
+						"port": 1981,
+						"weight": 1
+					},
+					{
+						"host": "172.16.238.20",
+						"port": 1982,
+						"weight": 1
+					}],
+					"type": "chash",
+					"key": "query_string"
+				}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+		},
+		{
+			caseDesc: "create route using the upstream just created",
+			Object:   ManagerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/1",
+			Body: `{
+				"uri": "/server_port",
+				"upstream_id": "1"
+			}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+			Sleep:        sleepTime,
+		},
+	}
+
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+
+	//hit routes
+	time.Sleep(time.Duration(500) * time.Millisecond)
+	basepath := "http://127.0.0.1:9080"
+	var url string
+	var respBody []byte
+	res := map[string]int{}
+	for i := 0; i <= 17; i++ {
+		url = basepath + "/server_port?var=2&var2=" + strconv.Itoa(i)
+		req, err := http.NewRequest("GET", url, nil)
+		resp, err := http.DefaultClient.Do(req)
+		assert.Nil(t, err)
+		respBody, err = ioutil.ReadAll(resp.Body)
+		body := string(respBody)
+		if _, ok := res[body]; !ok {
+			res[body] = 1
+		} else {
+			res[body] += 1
+		}
+		resp.Body.Close()
+	}
+	//the results will be distributed among the 3 upstream
+	assert.Equal(t, 4, res["1980"])
+	assert.Equal(t, 9, res["1981"])
+	assert.Equal(t, 5, res["1982"])
+}
+
+func TestUpstream_chash_arg_xxx(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "create chash upstream with key (arg_xxx)",
+			Object:   ManagerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/upstreams/1",
+			Body: `{
+					"nodes": [{
+						"host": "172.16.238.20",
+						"port": 1980,
+						"weight": 1
+					},
+					{
+						"host": "172.16.238.20",
+						"port": 1981,
+						"weight": 1
+					},
+					{
+						"host": "172.16.238.20",
+						"port": 1982,
+						"weight": 1
+					}],
+					"type": "chash",
+					"key": "arg_device_id"
+				}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+		},
+		{
+			caseDesc: "create route using the upstream just created",
+			Object:   ManagerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/1",
+			Body: `{
+				"uri": "/server_port",
+				"upstream_id": "1"
+			}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+			Sleep:        sleepTime,
+		},
```

Comment:
https://github.com/apache/apisix-dashboard/pull/932/files#diff-13ef2604ee721f34005fc15685defc18abb4562ea1a2a95d7931e89b0f59c19dR132-R144

we can delete this step.

Reply:
fixed.

Reply:
fixed